src/gui/sdlmain.cpp

```diff
@@ -2216,47 +2287,119 @@ void GFX_Events() {
-					continue;
-				/*
-				 *  EXPOSED indicates that the window needs to be redrawn.
-				 *  Note that on Windows/Linux-X11/Wayland/macOS, the EXPOSED event
-				 *  is fired after toggling between full vs windowed modes. However this is
-				 *  never fired on the Raspberry Pi (when rendering to the Framebuffer);
-				 *  therefore we rely on the FOCUS_GAINED event to catch window startup
-				 *  and size toggles.
-				 */ 
-				case SDL_WINDOWEVENT_EXPOSED:
-					if (sdl.draw.callback)
-						sdl.draw.callback(GFX_CallBackRedraw);
-					GFX_UpdateMouseState();
-					continue;
-				case SDL_WINDOWEVENT_FOCUS_GAINED:
-					SetPriority(sdl.priority.focus);
-					CPU_Disable_SkipAutoAdjust();
-					GFX_UpdateMouseState();
-					break;
-				case SDL_WINDOWEVENT_FOCUS_LOST:
+			case SDL_WINDOWEVENT_RESTORED:
+				DEBUG_LOG_MSG("SDL: Window has been restored");
+				/* We may need to re-create a texture
+				 * and more on Android. Another case:
+				 * Update surface while using X11.
+				 */
+				GFX_ResetScreen();
+				continue;
+
+			case SDL_WINDOWEVENT_RESIZED:
+				DEBUG_LOG_MSG("SDL: Window has been resized to %dx%d",
+				              event.window.data1,
+				              event.window.data2);
+				HandleVideoResize(event.window.data1,
+				                  event.window.data2);
+				continue;
+
+			case SDL_WINDOWEVENT_EXPOSED:
+				DEBUG_LOG_MSG("SDL: Window has been exposed "
+				              "and should be redrawn");
+				/* FIXME: below is not consistently true :(
+				 * seems incorrect on KDE and sometimes on MATE
+				 *
+				 * Note that on Windows/Linux-X11/Wayland/macOS,
+				 * event is fired after toggling between full vs
+				 * windowed modes. However this is never fired
+				 * on the Raspberry Pi (when rendering to the
+				 * Framebuffer); therefore we rely on the
+				 * FOCUS_GAINED event to catch window startup
+				 * and size toggles.
+				 */
+				GFX_ResetScreen();
+				GFX_UpdateMouseState();
+				continue;
+
+			case SDL_WINDOWEVENT_FOCUS_GAINED:
+				DEBUG_LOG_MSG("SDL: Window has gained keyboard focus");
+				SetPriority(sdl.priority.focus);
+				CPU_Disable_SkipAutoAdjust();
+				GFX_UpdateMouseState();
+				continue;
+
+			case SDL_WINDOWEVENT_FOCUS_LOST:
+				DEBUG_LOG_MSG("SDL: Window has lost keyboard focus");
 #ifdef WIN32
-					if (sdl.desktop.fullscreen) {
-						VGA_KillDrawing();
-						GFX_ForceFullscreenExit();
-					}
+				if (sdl.desktop.fullscreen) {
+					VGA_KillDrawing();
+					GFX_ForceFullscreenExit();
+				}
 #endif
-					SetPriority(sdl.priority.nofocus);
-					GFX_LosingFocus();
-					CPU_Enable_SkipAutoAdjust();
-					sdl.mouse.has_focus = false;
-					break;
-				default: ;
+				SetPriority(sdl.priority.nofocus);
+				GFX_LosingFocus();
+				CPU_Enable_SkipAutoAdjust();
+				sdl.mouse.has_focus = false;
+				break;
+
+			case SDL_WINDOWEVENT_ENTER:
+				DEBUG_LOG_MSG("SDL: Window has gained mouse focus");
+				continue;
```

Comment:
This block of events will be handy to help us find reliable triggers to use across platforms!
(and I'm guessing they will be elided away in optimized `-NDEBUG` builds).

Reply:
I am not sure what compilers will decide to do, perhaps they'll decide not to elide it, but it might be a *good* thing.

Notice how original code in here used `continue` (for enclosing while loop) instead of `break` (for switch) (!!!). This took me by surprise, because it's really easy to overlook. I decided to stick to the way original code in here worked (so still use continue, except few cases where break is used for the pause-related implementation few lines below).